executor/shuffle_merge_join.go

```diff
@@ -0,0 +1,363 @@
+package executor
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/execdetails"
+)
+
+// ShuffleMergeJoinExec implements the shuffle version of merge join algorithm
+type ShuffleMergeJoinExec struct {
+	baseExecutor
+	concurrency int
+	workers     []*shuffleMergeJoinWorker
+
+	prepared bool
+	executed bool
+
+	innerSplitter   partitionSplitter
+	outerSplitter   partitionSplitter
+	innerDataSource Executor
+	outerDataSource Executor
```

Comment:
It's better to support the outer data source on the original Shuffle instead of introducing a new struct.